File: apps/website/src/syncScheduler.ts
/**
 * Ordonnanceur de la synchronisation automatique (shared/spec/vault-sync.md).
 *
 * Pur : ni stockage ni réseau, seulement l'ordonnancement. `run` fait la
 * synchronisation, `canSync` dit si elle peut partir (session et clef
 * maîtresse). Minuteries et horloge sont injectables pour les tests.
 *
 * - les déclencheurs rapprochés sont fusionnés : attente de `delay` après le
 *   dernier ;
 * - une synchronisation en cours n'est jamais doublée : une demande pendant
 *   qu'elle tourne en relance une seule, juste après ;
 * - à l'ouverture, pas plus d'une fois toutes les `openThrottle` ms.
 */

export const AUTO_SYNC_DELAY_MS = 2000;
export const AUTO_SYNC_OPEN_THROTTLE_MS = 30000;

/** Issue d'une synchronisation : son résultat, un échec, ou rien de lancé. */
export type SyncRun<R> = { ok: true; value: R } | { ok: false; error: unknown } | { skipped: true };

export interface SyncSchedulerOptions<R> {
  run: () => Promise<R>;
  canSync?: () => boolean | Promise<boolean>;
  onResult?: (outcome: SyncRun<R>) => void;
  delay?: number;
  openThrottle?: number;
  now?: () => number;
  setTimer?: (fn: () => void, ms: number) => unknown;
  clearTimer?: (id: unknown) => void;
}

export interface SyncScheduler<R> {
  /** Après une écriture locale ou un réglage : attend `delay` de calme. */
  trigger(): void;
  /** À l'ouverture : espacé de `openThrottle`. Rend faux si ignoré. */
  triggerOpen(): boolean;
  /** Bouton « Synchroniser » : tout de suite, après celle en cours. */
  runNow(): Promise<SyncRun<R>>;
  /** Oublie ce qui attend (la synchronisation en cours va à son terme). */
  cancel(): void;
  readonly pending: boolean;
  readonly running: boolean;
}

export function createSyncScheduler<R>({
  run,
  canSync = () => true,
  onResult = () => {},
  delay = AUTO_SYNC_DELAY_MS,
  openThrottle = AUTO_SYNC_OPEN_THROTTLE_MS,
  now = () => Date.now(),
  setTimer = (fn, ms) => setTimeout(fn, ms),
  clearTimer = (id) => clearTimeout(id as ReturnType<typeof setTimeout>),
}: SyncSchedulerOptions<R>): SyncScheduler<R> {
  let timer: unknown = null;
  let running: Promise<SyncRun<R>> | null = null;
  let queued = false;
  let lastOpen: number | null = null;

  function cancelTimer() {
    if (timer !== null) clearTimer(timer);
    timer = null;
  }

  function execute(): Promise<SyncRun<R>> {
    const current = (async (): Promise<SyncRun<R>> => {
      let allowed = false;
      try {
        allowed = await canSync();
      } catch {
        allowed = false;
      }
      if (!allowed) return { skipped: true };
      let outcome: SyncRun<R>;
      try {
        outcome = { ok: true, value: await run() };
      } catch (error) {
        outcome = { ok: false, error };
      }
      try {
        onResult(outcome);
      } catch {
        // Le statut est un confort : son échec ne change pas le résultat.
      }
      return outcome;
    })();
    running = current;
    void current.then(() => {
      running = null;
      if (queued) {
        queued = false;
        void execute();
      }
    });
    return current;
  }

  function fire() {
    timer = null;
    if (running) {
      queued = true;
      return;
    }
    void execute();
  }

  function trigger() {
    cancelTimer();
    timer = setTimer(fire, delay);
  }

  function triggerOpen() {
    const t = now();
    if (lastOpen !== null && t - lastOpen < openThrottle) return false;
    lastOpen = t;
    trigger();
    return true;
  }

  async function runNow() {
    cancelTimer();
    queued = false;
    while (running) await running;
    return execute();
  }

  return {
    trigger,
    triggerOpen,
    runNow,
    cancel() {
      cancelTimer();
      queued = false;
    },
    get pending() {
      return timer !== null;
    },
    get running() {
      return running !== null;
    },
  };
}

File: apps/android/app/src/main/java/fr/juliette/thecode/vault/SyncScheduler.java
package fr.juliette.thecode.vault;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.concurrent.Executor;

/**
 * Quand lancer la synchronisation automatique (shared/spec/vault-sync.md,
 * « Synchronisation automatique »).
 *
 * Sans Android : l'horloge, la minuterie et le fil d'exécution sont injectés,
 * ce qui rend les règles vérifiables en test.
 *
 * - les demandes rapprochées sont fusionnées : on attend {@link #DEBOUNCE_MS}
 *   après la dernière ;
 * - une synchronisation en cours n'est jamais doublée : une demande pendant
 *   qu'elle tourne en relance une seule après ;
 * - à l'ouverture, pas plus d'une fois toutes les {@link #OPEN_THROTTLE_MS} ;
 * - rien ne part tant que {@link #ready} est faux (pas de compte, pas de clef).
 */
public final class SyncScheduler {

    public static final long DEBOUNCE_MS = 2_000;
    public static final long OPEN_THROTTLE_MS = 30_000;

    /** Horloge en millisecondes. */
    public interface Clock {
        long now();
    }

    /** Minuterie : lance une tâche après un délai, annulable. */
    public interface Timer {
        @NonNull
        Cancellable schedule(@NonNull Runnable task, long delayMs);
    }

    public interface Cancellable {
        void cancel();
    }

    /** Condition de départ : un compte lié et une clef maîtresse. */
    public interface Ready {
        boolean isReady();
    }

    private final Clock clock;
    private final Timer timer;
    private final Executor worker;
    private final Ready ready;
    private final Runnable task;

    @Nullable
    private Cancellable pending;
    private boolean running;
    private boolean rerun;
    private long lastOpenAt;
    private boolean opened;

    /**
     * @param task la synchronisation, exécutée sur {@code worker} ; ses échecs
     *             sont les siens, une exception n'arrête pas l'ordonnanceur
     */
    public SyncScheduler(@NonNull Clock clock, @NonNull Timer timer, @NonNull Executor worker,
                         @NonNull Ready ready, @NonNull Runnable task) {
        this.clock = clock;
        this.timer = timer;
        this.worker = worker;
        this.ready = ready;
        this.task = task;
    }

    /** Ouverture ou retour au premier plan : espacé de 30 secondes. */
    public synchronized void onOpen() {
        if (!ready.isReady()) return;
        long now = clock.now();
        if (opened && now - lastOpenAt < OPEN_THROTTLE_MS) return;
        opened = true;
        lastOpenAt = now;
        debounce();
    }

    /** Écriture locale ou réglage modifié : regroupé avec les suivants. */
    public synchronized void onChange() {
        if (!ready.isReady()) return;
        debounce();
    }

    /**
     * Toute écriture locale du carnet ({@link Vault#save}) relance la
     * synchronisation ; celles de la synchronisation elle-même
     * ({@link Vault#saveSynced}) non.
     */
    public void watchVaultWrites() {
        Vault.setWriteListener(this::onChange);
    }

    /** Bouton « Synchroniser » : tout de suite, sans jamais doubler une en cours. */
    public synchronized void runNow() {
        if (!ready.isReady()) return;
        cancelPending();
        start();
    }

    /** Vrai tant qu'une synchronisation tourne. */
    public synchronized boolean isRunning() {
        return running;
    }

    private void debounce() {
        cancelPending();
        pending = timer.schedule(this::fire, DEBOUNCE_MS);
    }

    private void cancelPending() {
        if (pending != null) {
            pending.cancel();
            pending = null;
        }
    }

    private synchronized void fire() {
        pending = null;
        start();
    }

    private void start() {
        if (running) {
            rerun = true;
            return;
        }
        running = true;
        worker.execute(() -> {
            try {
                if (ready.isReady()) task.run();
            } finally {
                finished();
            }
        });
    }

    private synchronized void finished() {
        running = false;
        if (rerun) {
            rerun = false;
            start();
        }
    }
}
